# Networks: structure, evolution & processes
**Internet Analytics - Lab 2**

---

**Group:** K

**Names:**

*  Xavier Jeanmonod
*  Adrian Baudat
*  Simon Wicky

---

#### Instructions

*This is a template for part 1 of the lab. Clearly write your answers, comments and interpretations in Markodown cells. Don't forget that you can add $\LaTeX$ equations in these cells. Feel free to add or remove any cell.*

*Please properly comment your code. Code readability will be considered for grading. To avoid long cells of codes in the notebook, you can also embed long python functions and classes in a separate module. Don’t forget to hand in your module if that is the case. In multiple exercises, you are required to come up with your own method to solve various problems. Be creative and clearly motivate and explain your methods. Creativity and clarity will be considered for grading.*

---

## 2.1 Exploration of real networks

#### Exercise 2.1

In [2]:
import networkx as nx
import numpy as npy

In [3]:
Data = open('../data/network1.csv', "r")
next(Data,None)
Graphtype = nx.Graph()
G = nx.parse_edgelist(Data, delimiter=',', create_using=Graphtype,
                      nodetype=int)
print("Number of edges : " + str(len(G.edges)))
print("Number of nodes : " + str(len(G.nodes)))

Number of edges : 18584
Number of nodes : 13033


#### Exercise 2.2:  Node degree distribution

#### Exercise 2.3: Giant component

In [4]:

print("Number of connect components : " + str(nx.number_connected_components(G)))
size = [len(o) for o in nx.connected_components(G)]

print("Histogram of connected components (size : number of CC)")
for i in range(max(size) + 1):
    if size.count(i) > 0:
        print(str(i) + " : " + str(size.count(i)))
    

Number of connect components : 192
Histogram of connected components (size : number of CC)
2 : 149
3 : 29
4 : 7
5 : 3
6 : 1
7 : 1
16 : 1
12576 : 1


Looking at the data above, we can see that there is 192 connected components, with one giant components of size 12756

#### Exercise 2.4: Short paths & Small-worlds

In [21]:
from random import choice
giant_cc = [e for e in nx.connected_component_subgraphs(G) if len(e.nodes()) == max(size)][0]
sum = 0
#iterations = 1000
iterations = 1
for n in range(iterations):
    n1 = choice(list(giant_cc.nodes))                
    for n2 in giant_cc.nodes:
        if n1 != n2:
            length = nx.shortest_path_length(giant_cc, source=n1, target=n2)
            sum += length
print(sum / len(giant_cc.nodes()) / iterations )

0
100
200
300
400
500
600
700
800
900
4.174713899491094


The average shortest path is computed as follow :
*    Pick one node at random, and computes its average shortest path length with every other node
*    Do step 1 on 1000 nodes and average the whole

This computation seems relatively representative of the average shortest path length, the real computation is too costly.

And so this computation yields 4.174

#### Exercise 2.5: Network comparison

5

#### Exercise 2.6: Network identification